docs/architecture/adr-063-privval-grpc.md

```diff
@@ -0,0 +1,75 @@
+
+Tendermint uses its own custom secure connection protocol (`SecretConnection`) and a raw tcp/unix socket connection protocol. The secure connection protocol until recently was exposed to man in the middle attacks and limits the amount and speed of integrations users can make. The raw tcp connection protocol is less custom, but has been causing minuet issues with users. 
+
+Migrating Tendermint's private validator to a widely adopted protocol will ease the current maintenance and integration burden experienced with the current protocol. 
+
+## Decision
+
+After discussing with multiple stake holders, [gRPC](https://grpc.io/) was decided on to replace the current private validator protocol. GRPC is a widely adopted protocol in the micro-service and cloud infrastructure world. GRPC uses [protocol-buffers](https://developers.google.com/protocol-buffers) to describe its services, providing a language agnostic implementation. Tendermint uses protobuf for on disk and over the wire encoding already making the integration with gRPC simpler. 
+
+## Alternative Approaches
+
+- JSON-RPC: We did not consider JSON-RPC because Tendermint uses protobuf extensively making gRPC a natural choice.
+
+## Detailed Design
+
+With the recent integration of [Protobuf](https://developers.google.com/protocol-buffers) into Tendermint the needed changes to migrate from the current private validator protocol to gRPC is not large. 
+
+The [service definition](https://grpc.io/docs/what-is-grpc/core-concepts/#service-definition) for gRPC will be defined as:
+
+```proto
+service PrivValidatorAPI {
+  rpc GetPubKey(tendermint.proto.privval.PubKeyRequest) returns (tendermint.proto.privval.PubKeyResponse);
+  rpc SignVote(tendermint.proto.privval.SignVoteRequest) returns (tendermint.proto.privval.SignedVoteResponse);
+  rpc SignProposal(tendermint.proto.privval.SignProposalRequest) returns (tendermint.proto.privval.SignedProposalResponse);
+}
+```
+
+#### Keep Alive
+
+If you have worked on the private validator system you will see that we are removing the `PingRequest` and `PingResponse` messages. These messages were used to create functionality which kept the connection alive. With gRPC there is a [keep alive feature](https://github.com/grpc/grpc/blob/master/doc/keepalive.md) that will be added along side the integration to provide the same functionality. 
+
+#### Metrics
+
+Remote signers are crucial to operating secure and consistently up Validators. In the past there were no metrics to tell the operator if something is wrong other than the node not signing. Integrating metrics into the client and provided server will be done with [prometheus](https://github.com/grpc-ecosystem/go-grpc-prometheus). This will be integrated into node's prometheus export for node operators. 
```

Comment:
Metrics are registered on the server side (the remote private validator) right? or also on the Tendermint side? Do we want to provide support for multiple metric providers like we do with the actual nodes? or just standardise Prometheus

Reply:
> Metrics are registered on the server side (the remote private validator) right? or also on the Tendermint side?

Tendermint side as well

> Do we want to provide support for multiple metric providers like we do with the actual nodes? or just standardise Prometheus

Tendermint only has Prometheus not sure what you mean by multiple metric providers?

Reply:
> Tendermint only has Prometheus not sure what you mean by multiple metric providers?

I see. I was put off by something in the code that made me think we supported multiple. 

> Tendermint side as well

and we already know what metrics we want to register in this regard right?

Reply:
the metrics documentation will point to: https://github.com/grpc-ecosystem/go-grpc-prometheus. It has a list of already created metrics and if we get requests for more we can add custom ones.